paddle/fluid/operators/conv_mkldnn_op.cc

```diff
@@ -358,132 +364,424 @@ class ConvMKLDNNOpKernel : public paddle::framework::OpKernel<T> {
+          (g == 1) ? mkldnn::memory::format::oihw
+                   : mkldnn::memory::format::goihw);
+      /* create memory descriptor for convolution without specified format
+       * ('any') which lets a primitive (convolution in this case) choose
+       * the memory format preferred for best performance
+      */
+      std::string data_format = ctx.Attr<std::string>("data_format");
+      auto chosen_memory_format =
+          platform::data_format_to_memory_format(data_format);
+
+      auto bias_tz = paddle::framework::vectorize2int(bias->dims());
+
+      auto src_md = platform::MKLDNNMemDesc(src_tz, memory::data_type::u8,
+                                            chosen_memory_format);
+      auto weights_md = platform::MKLDNNMemDesc(
+          weights_tz, memory::data_type::s8, chosen_memory_format);
+
+      auto dst_dt = fuse_relu ? paddle::framework::ToMKLDNNDataType(
+                                    std::type_index(typeid(unsigned char)))
+                              : paddle::framework::ToMKLDNNDataType(
+                                    std::type_index(typeid(signed char)));
+
+      if (fuse_residual_conn) {
+        auto residual = ctx.Input<Tensor>("ResidualData");
+        auto residual_dt =
+            paddle::framework::ToMKLDNNDataType(residual->type());
+        if (dst_dt != residual_dt) dst_dt = residual_dt;
+      }
+      auto dst_md =
+          platform::MKLDNNMemDesc(dst_tz, dst_dt, chosen_memory_format);
+
+      // create a conv primitive descriptor and save it for usage in backward
+      std::shared_ptr<mkldnn::convolution_forward::primitive_desc> conv_pd;
+
+      if (bias) {
+        auto bias_md = platform::MKLDNNMemDesc(bias_tz, memory::data_type::s32,
+                                               memory::format::x);
+        conv_pd = ConvFwdPrimitiveDesc(
+            src_md, weights_md, bias_md, dst_md, strides, paddings,
+            mkldnn_engine, fuse_relu, fuse_residual_conn, output_shift_scale,
+            sum_scale[0], is_test);
+      } else {
+        conv_pd =
+            ConvFwdPrimitiveDesc(src_md, weights_md, dst_md, strides, paddings,
+                                 mkldnn_engine, fuse_relu, fuse_residual_conn,
+                                 output_shift_scale, sum_scale[0], is_test);
+      }
+      // Save conv_pd/src_memory/weights_memory for backward pass
+      dev_ctx.SetBlob(key_conv_pd, conv_pd);
+
+      ConvMKLDNNHandler handler(conv_pd, dev_ctx, mkldnn_engine, key);
+
+      // create mkldnn memory from input tensors (data/weights)
+      auto user_src_memory_p =
+          handler.AcquireSrcMemory(user_src_md, to_void_cast<T>(input_data));
+      auto user_weights_memory_p = handler.AcquireWeightsMemory(
+          user_weights_md, to_void_cast<float>(filter_data));
+
+      // create reorder primitive if the input format is not the preferred one
+      auto src_memory_p =
+          handler.AcquireSrcMemoryFromPrimitive(user_src_memory_p, pipeline);
+
```

Comment:
It seems that code in lines 571 to 638 is similar to code already present in if-block of this if-else statement. Do you think it could be refactored somehow so it's not repeated?

Reply:
yes, we are planning to restructure the code on next PR, and keep these code for current PR, we want to commit it as soon as possible because calibration tool check also needs it.

Reply:
Synced with Poland team, we agreed to have such option in current implementation. @Sand3r- @tpatejko @kbinias